File: internal/limiter/token_bucket.go
package limiter

import (
	"context"
	_ "embed"
	"fmt"
	"time"

	"github.com/redis/go-redis/v9"
)

//go:embed lua/token_bucket.lua
var luaTokenBucketScript string

// TokenBucketConfig holds Token Bucket algorithm parameters.
type TokenBucketConfig struct {
	Capacity   int
	RefillRate float64
	KeyTTL     int
}

// DefaultTokenBucketConfig returns sensible defaults: 100 req/min = ~1.67 tokens/sec
func DefaultTokenBucketConfig() TokenBucketConfig {
	return TokenBucketConfig{
		Capacity:   100,
		RefillRate: 100.0 / 60.0,
		KeyTTL:     120,
	}
}

// TokenBucket implements Token Bucket rate limiting using Redis.
type TokenBucketLimiter struct {
	client *redis.Client
	config TokenBucketConfig
	script *redis.Script
}

// NewTokenBucket creates a new Token Bucket limiter.
func NewTokenBucket(client *redis.Client, config TokenBucketConfig) *TokenBucketLimiter {
	if config.Capacity <= 0 {
		config.Capacity = 100
	}
	if config.RefillRate <= 0 {
		config.RefillRate = 100.0 / 60.0
	}
	if config.KeyTTL <= 0 {
		config.KeyTTL = 120
	}
	return &TokenBucketLimiter{
		client: client,
		config: config,
		script: redis.NewScript(luaTokenBucketScript),
	}
}

func (tb *TokenBucketLimiter) Allow(ctx context.Context, key string) (bool, error) {
	fullKey := fmt.Sprintf("rate_limit:%s", key)
	now := float64(time.Now().UnixMicro()) / 1e6

	result, err := tb.script.Run(ctx, tb.client, []string{fullKey},
		tb.config.Capacity,
		tb.config.RefillRate,
		now,
		1,
		tb.config.KeyTTL,
	).Int()
	if err != nil {
		return false, err
	}
	return result == 1, nil
}

File: internal/limiter/lua/token_bucket.lua
-- Token Bucket rate limiter - atomic Redis implementation
-- KEYS[1]: rate limit key (e.g., "rate_limit:192.168.1.1")
-- ARGV[1]: capacity (max tokens)
-- ARGV[2]: refill rate (tokens per second)
-- ARGV[3]: current timestamp (Unix seconds)
-- ARGV[4]: tokens to consume (default 1)
-- ARGV[5]: key TTL in seconds

local key = KEYS[1]
local capacity = tonumber(ARGV[1])
local refill_rate = tonumber(ARGV[2])
local now = tonumber(ARGV[3])
local requested = tonumber(ARGV[4]) or 1
local ttl = tonumber(ARGV[5]) or 120

local data = redis.call('HMGET', key, 'tokens', 'last_update')
local tokens = tonumber(data[1]) 
local last_update = tonumber(data[2])

if tokens == nil then
    tokens = capacity
    last_update = now
end

-- Refill tokens based on elapsed time
local elapsed = now - last_update
local refill = elapsed * refill_rate
tokens = math.min(capacity, tokens + refill)

if tokens >= requested then
    tokens = tokens - requested
    redis.call('HMSET', key, 'tokens', tokens, 'last_update', now)
    redis.call('EXPIRE', key, ttl)
    return 1 -- allowed
else
    return 0 -- denied
end